qiskit_nature/operators/second_quantization/spin_op.py

```diff
@@ -18,111 +18,497 @@
+            label_list = label.split()
+            for single in label_list:
+                if single[0] not in xyz_dict:
+                    raise ValueError(f"Given label {single} must be X, Y, Z, or I.")
+                xyz_num = xyz_dict[single[0]]
+                match_index = re_index.search(single)
+                if match_index:
+                    index = int(match_index.group())
+                else:
+                    raise ValueError(f"Given label {single} has no index.")
+                if len(single) == match_index.end():
+                    power = 1
+                elif (
+                        single[match_index.end()] == "^"
+                        and single[match_index.end() + 1:].isdecimal()
+                ):
+                    power = int(single[match_index.end() + 1:])
+                else:
+                    raise ValueError(f"Invalid label: {single}.")
+
+                max_index = max(max_index, index)
+
+                if xyz_num != 0:
+                    parsed_data_term.append((xyz_num, index, power))
+            parsed_data.append(parsed_data_term)
+
+        self._register_length = max_index + 1
+        self._spin_array = np.zeros((num_terms, self._register_length * 3), dtype=np.uint8)
+        for i, data in enumerate(parsed_data):
+            for datum in data:
+                if (
+                        datum[0] == 1
+                        and self._spin_array[i][2 * self._register_length - datum[1] - 1] > 0
+                ):
+                    raise ValueError("Label must be XYZ order.")
+                if (
+                        datum[0] == 1
+                        and self._spin_array[i][3 * self._register_length - datum[1] - 1] > 0
+                ):
+                    raise ValueError("Label must be XYZ order.")
+                if (
+                        datum[0] == 2
+                        and self._spin_array[i][3 * self._register_length - datum[1] - 1] > 0
+                ):
+                    raise ValueError("Label must be XYZ order.")
+
+                if self._spin_array[i][datum[0] * self._register_length - datum[1] - 1] != 0:
+                    raise ValueError("Duplicate label.")
+                self._spin_array[i][datum[0] * self._register_length - datum[1] - 1] = datum[2]
+
+    def _parse_dense_label(self, labels):
+        self._register_length = len(labels[0])
+        num_terms = len(labels)
+        self._spin_array = np.zeros((num_terms, self._register_length * 3), dtype=np.uint8)
+        for i, label in enumerate(labels):
+            for pos, char in enumerate(label):
+                if char == "X":
+                    self._spin_array[i][pos] = 1
+                elif char == "Y":
+                    self._spin_array[i][self._register_length + pos] = 1
+                elif char == "Z":
+                    self._spin_array[i][2 * self._register_length + pos] = 1
+
+    def _parse_ladder(self, data):
```

Comment:
Please ensure to document somewhere, that this method *must* be called before parsing a sparse or dense label because those respective methods cannot handle ladder operators themselves. It may even be a good idea to assert that neither label contains a `+` or `-` in the beginning of the actual parse methods.